File: Graphs/AdjacencyMatrixWeight.py
class Graph_Weight:
    def __init__(self, size):
        self.size = size
        self.matrix = [[None] * size for _ in range(size)]  
        self.vertex_data = [" "] * size  # List to store vertex data
    
    def addEdge(self, i, j, weight):
        if 0 <= i < self.size and 0 <= j < self.size and weight > 0:
            self.matrix[i][j] = weight
    
    def addVertex(self, vertex, data):
        if 0 <= vertex < self.size:
            self.vertex_data[vertex] = data

    def dfs_util(self, v, visited):
        visited[v] = True
        print(self.vertex_data[v], end=" ")

        for i in range(self.size):
            if self.matrix[v][i] is not None and not visited[i]:  
                self.dfs_util(i, visited)
    
    def dfs(self, startVertexData):
        visited = [False] * self.size
        startVertex = self.vertex_data.index(startVertexData)
        self.dfs_util(startVertex, visited)

    def bfs(self, startVertexData):
        queue = [self.vertex_data.index(startVertexData)]
        visited = [False] * self.size
        visited[queue[0]] = True

        while queue:
            current_vertex = queue.pop(0)
            print(self.vertex_data[current_vertex], end=' ')
            for i in range(self.size):
                if self.matrix[current_vertex][i] is not None and not visited[i]:  
                    queue.append(i)
                    visited[i] = True
    
    def dijkstra(self, startVertexData):
        startVertex = self.vertex_data.index(startVertexData)
        distance = [float('inf')] * self.size
        distance[startVertex] = 0
        visited = [False] * self.size

        for _ in range(self.size):
            min_distance = float('inf')
            u = None
            for i in range(self.size):
                if not visited[i] and distance[i] < min_distance:
                    min_distance = distance[i]
                    u = i

            if u is None:
                break

            visited[u] = True

            for v in range(self.size):
                if self.matrix[u][v] is not None and not visited[v]:
                    alt = distance[u] + self.matrix[u][v]
                    if alt < distance[v]:
                        distance[v] = alt

        return distance
    
    def singleDestinationVertex(self, startVertexData, endVertexData):
        startVertex = self.vertex_data.index(startVertexData)
        endVertex = self.vertex_data.index(endVertexData)
        
        distances = [float('inf')] * self.size
        predecessors = [None] * self.size
        distances[startVertex] = 0
        visited = [False] * self.size

        for _ in range(self.size):
            min_distance = float('inf')
            u = None
            for i in range(self.size):
                if not visited[i] and distances[i] < min_distance:
                    min_distance = distances[i]
                    u = i
            
            if u is None or u == endVertex:
                break

            visited[u] = True

            for v in range(self.size):
                if self.matrix[u][v] is not None and not visited[v]: 
                    alt = distances[u] + self.matrix[u][v]
                    if alt < distances[v]:
                        distances[v] = alt
                        predecessors[v] = u
        
        path = []
        current = endVertex
        while current is not None:
            path.insert(0, self.vertex_data[current])
            current = predecessors[current]

        return distances[endVertex], path
    
    def printGraph(self):
        print("Adjacency Matrix:")
        for row in self.matrix:
            print(" ".join(map(lambda x: str(x) if x is not None else "0", row)))  
        print("\nVertex Data:")
        for vertex, data in enumerate(self.vertex_data):
            print(f"Vertex {vertex}: {data}")

File: Graphs/AdjacencyMatrixWeight.java
package DSA;

import java.util.Arrays;
import java.util.LinkedList;
import java.util.Queue;

public class Graphs_Weight {
    int size;
    int[][] matrix;
    String[] vertex_data;
    public Graphs_Weight(int size) {
        this.size = size;
        this.matrix = new int[size][size];
        this.vertex_data = new String[size];
    }
    public void addEdge(int i, int j, int weight) {
        if ((i >= 0 && i < size) && (j >= 0 && j < size)) {
            if (weight > 0)
                matrix[i][j] = weight;
            else
                System.out.println("Trọng số phải là số dương.");
        }
    }
    public void addVertex(int vertex, String data) {
        if (vertex >= 0 && vertex < size) {
            vertex_data[vertex] = data;
        }
    }
    public void dfs_util(int v, boolean[] visited) {
        visited[v] = true;
        System.out.print(vertex_data[v] + " ");
        for (int i = 0; i < size; i++) {
            if (matrix[v][i] > 0 && !visited[i]) 
                dfs_util(i, visited);
            
        }
    }

    public void dfs(int startVertex) {
        boolean[] visited = new boolean[size];
        dfs_util(startVertex, visited);
    }

    
    public void bfs(int startVertex) {
        boolean[] visited = new boolean[size];
        Queue<Integer> queue = new LinkedList<>();
        queue.add(startVertex);
        visited[startVertex] = true;

        while (!queue.isEmpty()) {
            int currentVertex = queue.poll();
            System.out.print(vertex_data[currentVertex] + " ");
            for (int i = 0; i < size; i++) {
                if (matrix[currentVertex][i] > 0 && !visited[i]) {
                    queue.add(i);
                    visited[i] = true;
                }
            }
        }
    }

	
    public int getMinDistance(int[] dist, boolean[] visited) {
        int min = Integer.MAX_VALUE;
        int minIndex = 0;
        for (int i = 0; i < size; i++) {
            if (!visited[i] && dist[i] < min) {
                min = dist[i];
                minIndex = i;
            }
        }
        return minIndex;
    }
    public int dijkstra(String start_vertex_data) {
        int start_vertex = -1;
        for (int i = 0; i < vertex_data.length; i++) {
            if (vertex_data[i].equals(start_vertex_data)) {
                start_vertex = i;
                break; 
            }
        }
        if (start_vertex == -1) {
            System.out.println("Không tìm thấy đỉnh với dữ liệu: " + start_vertex_data);
            return -1;
        }
        int[] dist = new int[size];
        Arrays.fill(dist, Integer.MAX_VALUE); // inf 
        dist[start_vertex] = 0; // luon bat dau voi dinh = 0

        boolean[] visited = new boolean[size];

        //Duyet graphs
        for (int i = 0; i < size; i++) {
            int u = getMinDistance(dist, visited); 
            visited[u] = true;
            for (int v = 0; v < size; v++) {
                if (!visited[v] && matrix[u][v] > 0 && dist[u] != Integer.MAX_VALUE && dist[u] + matrix[u][v] < dist[v]) {
                    dist[v] = dist[u] + matrix[u][v];
                }
            }
        }
        System.out.println("Khoảng cách từ đỉnh " + start_vertex_data + " đến các đỉnh khác:");
        for (int i = 0; i < size; i++) {
            if (dist[i] == Integer.MAX_VALUE) {
                System.out.println(vertex_data[i] + ": Không thể đến");
            } else {
                System.out.println(vertex_data[i] + ": " + dist[i]);
            }
        }
        return 0;
    }
    public void print() {
    	 System.out.println("Dữ liệu các đỉnh:");
         for (int i = 0; i < size; i++) {
             System.out.println(vertex_data[i] + " : " + i);
         }
        System.out.println("Ma trận kề:");
        for (int i = 0; i < size; i++) {
            for (int j = 0; j < size; j++) {
                System.out.print(matrix[i][j] + " ");
            }
            System.out.println();
        }

       
    }
}
